presto-native-execution/presto_cpp/main/PeriodicServiceInventoryManager.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "presto_cpp/main/PeriodicServiceInventoryManager.h"
+#include <folly/futures/Retrying.h>
+#include <velox/common/memory/Memory.h>
+
+namespace facebook::presto {
+PeriodicServiceInventoryManager::PeriodicServiceInventoryManager(
+    std::string address,
+    int port,
+    std::shared_ptr<CoordinatorDiscoverer> coordinatorDiscoverer,
+    std::string clientCertAndKeyPath,
+    std::string ciphers,
+    std::string id,
+    uint64_t frequencyMs)
+    : address_(std::move(address)),
+      port_(port),
+      coordinatorDiscoverer_(std::move(coordinatorDiscoverer)),
+      clientCertAndKeyPath_(std::move(clientCertAndKeyPath)),
+      ciphers_(std::move(ciphers)),
+      id_(std::move(id)),
+      frequencyMs_(frequencyMs),
+      pool_(velox::memory::addDefaultLeafMemoryPool(id_)),
+      eventBaseThread_(false /*autostart*/) {}
+
+void PeriodicServiceInventoryManager::start() {
+  eventBaseThread_.start(id_);
+  stopped_ = false;
+  auto* eventBase = eventBaseThread_.getEventBase();
+  eventBase->runOnDestruction([this] { client_.reset(); });
+  eventBase->schedule([this]() { return sendRequest(); });
+}
+
+void PeriodicServiceInventoryManager::stop() {
+  stopped_ = true;
+  eventBaseThread_.stop();
+}
+
+void PeriodicServiceInventoryManager::sendRequest() {
+  // stop() calls EventBase's destructor which executed all pending callbacks;
+  // make sure not to do anything if that's the case
+  if (stopped_) {
+    return;
+  }
+
+  try {
+    folly::SocketAddress newAddress = coordinatorDiscoverer_->getAddress();
+    if (updateServiceTimes() > 0 && (attempts_++) % updateServiceTimes() == 0) {
+      newAddress = coordinatorDiscoverer_->updateAddress();
+    }
```

Comment:
Please describe the logic here with some comments. Previous logic was very trivial. If newAddress != currentAddress, currentAddress was updated. Not clear about updateServiceTime().